backup/tar.go

```diff
@@ -0,0 +1,144 @@
+package backup
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/square/keysync/output"
+
+	"github.com/pkg/errors"
+)
+
+// Given a path to a directory, create and return a tarball of its content.
+// Careful, as this will pull the full contents into memory.
+// This is not a general-purpose function, but is intended to only work with Keysync
+// directories, which contain only non-executable regular files.
+func createTar(dir string) ([]byte, error) {
+	var tarball bytes.Buffer
+	tw := tar.NewWriter(&tarball)
+
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() || !info.Mode().IsRegular() {
+			// Skip directories and non-regular files.
+			return nil
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close() // We explicitly call close below with error handling, but this extra one handles early returns
+
+		// 2nd Argument to FileInfoHeader is only used for symlinks, which aren't relevant here
+		header, err := tar.FileInfoHeader(info, "")
+		if err != nil {
+			return err
+		}
+
+		// Set the name to be relative to the base directory
+		header.Name, err = filepath.Rel(dir, path)
+		if err != nil {
+			return err
+		}
+
+		if err := tw.WriteHeader(header); err != nil {
+			return err
+		}
+
+		if _, err := io.Copy(tw, f); err != nil {
+			return err
+		}
+
+		if err := f.Close(); err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := tw.Close(); err != nil {
```

Comment:
Does this close() also need to be deferred?  (Looking online deferring the close seems to be a common pattern, though the official docs don't warn either way that I could find.)

Reply:
I looked at this a bit more and added a comment:

	// Tar writing adds a trailer in Close(), and can possibly return errors, so we need to check
	// errors here.  We could also defer tw.Close(), but there's nothing to leak in the Tar Writer
	// other than the io.Writer that's passed in.  That's the tarball buffer in this function, so
	// we don't need to worry about leaking FDs. Calling Close() a 2nd time is always an error, so
	// I think it makes the error handling trickier if we both explicitly and defer a call to Close.

Reply:
Sounds good!  Thanks for looking into it.